src/scalems/radical/runtime.py

```diff
@@ -1042,20 +1042,34 @@ def runtime_shutdown(runtime: Runtime):
             logger.error("Runtime Session is already closed?!")
         else:
             if runtime.scheduler is not None:
-                # Cancel the master.
-                logger.debug("Canceling the master scheduling task.")
-                task_manager = runtime.task_manager()
-                task_manager.cancel_tasks(uids=runtime.scheduler.uid)
-                # Cancel blocks until the task is done so the following wait is
-                # (currently) redundant, but there is a ticket open to change this
-                # behavior.
-                # See https://github.com/radical-cybertools/radical.pilot/issues/2336
-                runtime.scheduler.wait(state=rp.FINAL)
-                logger.info(f"Master scheduling task complete: {repr(runtime.scheduler)}.")
+                # Note: The __aexit__ for the RuntimeManager makes sure that a `stop`
+                # is issued after the work queue is drained, if the scheduler task has
+                # not already ended.
+                runtime.scheduler.wait(rp.FINAL, timeout=10)
+                if runtime.scheduler.state not in rp.FINAL:
+                    # Cancel the master.
+                    logger.debug("Canceling the master scheduling task.")
+                    # Note: the effect of CANCEL is to send SIGTERM to the shell that
+                    # launched the task process.
+                    # TODO: Report incomplete tasks and handle the consequences of terminating the
+                    #  Raptor processes early.
+                    task_manager = runtime.task_manager()
+                    task_manager.cancel_tasks(uids=runtime.scheduler.uid)
+                    # Cancel blocks until the task is done so the following wait is
+                    # (currently) redundant, but there is a ticket open to change this
+                    # behavior.
+                    # See https://github.com/radical-cybertools/radical.pilot/issues/2336
+                # TODO(#249): wrap in a thread.
+                runtime.scheduler.wait(state=rp.FINAL, timeout=60)
```

Comment:
```suggestion
                    task_manager.cancel_tasks(uids=runtime.scheduler.uid)
                # As of https://github.com/radical-cybertools/radical.pilot/pull/2702,
                # we do not expect `cancel` to block, so we must wait for the
                # cancellation to succeed. It shouldn't take long, but it is not
                # instantaneous or synchronous. We hope that a minute is enough.
                # TODO(#249): wrap in a thread, since this could take a few seconds.
                runtime.scheduler.wait(state=rp.FINAL, timeout=60)
```